File: code/ch10-01/generate_calls.erl
%% @author J D Eisenberg <jdavid.eisenberg@gmail.com>
%% @doc Generate a random set of data for phone calls
%% @copyright 2013 J D Eisenberg
%% @version 0.1

-module(generate_calls).
-export([make_call_list/1, format_date/1, format_time/1]).

make_call_list(N) ->
  Now = calendar:datetime_to_gregorian_seconds({{2013, 3, 10}, {9, 0, 0}}),
  Numbers = [
    {"213-555-0172", Now},
    {"301-555-0433", Now},
    {"415-555-7871", Now},
    {"650-555-3326", Now},
    {"729-555-8855", Now},
    {"838-555-1099", Now},
    {"946-555-9760", Now}
  ],
  CallList = make_call_list(N, Numbers, []),
  {Result, OutputFile} = file:open("call_list.csv", [write]),
  case Result of
    ok -> write_item(OutputFile, CallList);
    error -> io:format("Error: ~p~n", OutputFile)
  end.

make_call_list(0, _Numbers, Result) -> lists:reverse(Result);

make_call_list(N, Numbers, Result) ->
  Entry = random:uniform(length(Numbers)),
  {Head, Tail} = lists:split(Entry - 1, Numbers),
  {Number, LastCall} = hd(Tail),
  StartCall = LastCall + random:uniform(120) + 20,
  Duration = random:uniform(180) + 40,
  EndCall = StartCall + Duration,
  Item = [Number, format_date(StartCall), format_time(StartCall),
    format_date(EndCall), format_time(EndCall)],
  UpdatedNumbers = Head ++ [{Number, EndCall} | tl(Tail)],
  make_call_list(N - 1, UpdatedNumbers, [Item | Result]).

write_item(OutputFile, []) ->
  file:close(OutputFile);

write_item(OutputFile, [H|T]) ->
  io:format("~s ~s ~s ~s ~s~n", H),
  io:fwrite(OutputFile, "~s,~s,~s,~s,~s~n", H),
  write_item(OutputFile, T).

format_date(GSeconds) ->
  {Date, _Time} = calendar:gregorian_seconds_to_datetime(GSeconds),
  {Y, M, D} = Date,
  lists:flatten(io_lib:format("~4b-~2..0b-~2..0b", [Y, M, D])).

format_time(GSeconds) ->
  {_Date, Time} = calendar:gregorian_seconds_to_datetime(GSeconds),
  {M, H, S} = Time,
  lists:flatten(io_lib:format("~2..0b:~2..0b:~2..0b", [M, H, S])).
  

File: code/ch10-01/generate_calls.ex
defmodule GenerateCalls do
  @moduledoc """ 
  Generate a random set of phone calls
  
  from *Études for Elixir*, O'Reilly Media, Inc., 2013.
  Copyright 2013 by J. David Eisenberg.
  """ 
  @vsn 0.1 

  def make_call_list(n) do
    now = :calendar.datetime_to_gregorian_seconds({{2013, 3, 10},
      {9, 0, 0}})
    numbers = [
      {"213-555-0172", now},
      {"301-555-0433", now},
      {"415-555-7871", now},
      {"650-555-3326", now},
      {"729-555-8855", now},
      {"838-555-1099", now},
      {"946-555-9760", now}
    ]
    call_list = make_call_list(n, numbers, [])
    {result, output_file} = File.open("call_list.csv", [:write,
      :utf8])
    case result do
      :ok -> write_item(output_file, call_list)
      :error -> IO.puts("Error creating output file")
    end
  end
  
  def make_call_list(0, _numbers, result) do
    Enum.reverse(result)
  end
  
  def make_call_list(n, numbers, result) do
    entry = :random.uniform(Enum.count(numbers))
    {head, tail} = Enum.split(numbers, entry - 1)
    {number, last_call} = hd(tail)
    start_call = last_call + :random.uniform(120) + 20
    duration = :random.uniform(180) + 40
    end_call = start_call + duration
    item = [number, format_date(start_call), format_time(start_call),
      format_date(end_call), format_time(end_call)]
    updated_numbers = head ++ [{number, end_call} | tl(tail)]
    make_call_list(n - 1, updated_numbers, [item | result])
  end
  
  def write_item(output_file, []) do
    File.close(output_file)
  end
  
  def write_item(output_file, [h | t]) do
    [n, sd, st, ed, et] = h
    IO.write(output_file, "#{n},#{sd},#{st},#{ed},#{et}\n")
    write_item(output_file, t)
  end
  
  def format_date(g_seconds) do
    {{y, m, d}, _time} =
      :calendar.gregorian_seconds_to_datetime(g_seconds)
    "#{y}-#{leadzero(m)}-#{leadzero(d)}"
  end
  
  def format_time(g_seconds) do
    {_date, {h, m, s}} =
      :calendar.gregorian_seconds_to_datetime(g_seconds)
    "#{leadzero(h)}:#{leadzero(m)}:#{leadzero(s)}"
  end
  
  def leadzero(n) do
    cond do
      n < 10 -> "0" <> integer_to_binary(n)
      true -> integer_to_binary(n)
    end
  end
end
